### DISCUSSION TITLE
Finite strain rotations

### DISCUSSION THREAD

**Discussion by tophmatthews** [idaholab/moose#16303] [Q&A Modules: Solid mechanics] [answered]
Hi All,

I have a question about finite strain rotations, specifically `perform_finite_strain_rotations` in `ADComputeMultipleInelasticStress`. When should `perform_finite_strain_rotations` be true (or conversely false). Trying to read around a bit, it looks like you don't need to do the rotations if everything is small strains...but what does that mean? Can I get away from the rotations if I am only dealing with things growing or shrinking, and not rotating?

I ask because first, it can be a bit expensive doing so many tensor operations in AD. ~~Also, and maybe more concerning, is the stress calculation in `ADComputeMultipleInelasticStress` vs "stress update" classes, such as `ADRadialReturnStressUpdate` derivatives. In `ADComputeMultipleInelasticStress`, the stress is calculated:~~

```
  if (_is_elasticity_tensor_guaranteed_isotropic || !_perform_finite_strain_rotations)
    _stress[_qp] = _elasticity_tensor[_qp] * (_elastic_strain_old[_qp] + elastic_strain_increment);
  else
    _stress[_qp] = _stress_old[_qp] + _elasticity_tensor[_qp] * elastic_strain_increment;
```

~~While in `ADRadialReturnStressUpdate`, the stress is calculated:~~
```
// Use the old elastic strain here because we require tensors used by this class
// to be isotropic and this method natively allows for changing in time
// elasticity tensors
stress_new = elasticity_tensor * (elastic_strain_old + strain_increment);
```

~~Alright, makes sense using the `ADRadialReturnStressUpdate` technique since I typically use an evolving elasticity tensor, but won't this lead to a mismatch in stress calculations when `perform_finite_strain_rotations=false`?~~

I guess the question is, when should I _not_ do `perform_finite_strain_rotations=false`. Maybe looking at the TM pros @bwspenc @sapitts @jiangwen84...

Thanks!

-Topher

**Comment by jiangwen84** [comment]
Since @recuero has been working on this recently, he might be able to give more details.

**Comment by recuero** [comment, marked as answer]
@tophmatthews My two cents below. I do not know much about the history of these classes though...

> Can I get away from the rotations if I am only dealing with things growing or shrinking, and not rotating?

Maybe. I believe you are suggesting mixing small and large strain kinematic descriptions, i.e. the rotation matrix being unity. This solution would require having prior knowledge that the system does not rotate in any part of it (not just rigid body rotations), which may happen in very particular cases.

> Alright, makes sense using the ADRadialReturnStressUpdate technique since I typically use an evolving elasticity tensor, but won't this lead to a mismatch in stress calculations when perform_finite_strain_rotations=false?

Doesn't `ADRadialReturnStressUpdate` override the calculation of stresses in `ADComputeMultipleInelasticStress`? I think I don't understand.

> when should I not do perform_finite_strain_rotations=false

I would choose `perform_finite_strain_rotations=true` any time you have large deformation kinematics. Those transformations appear necessary for that application. True, AD tensor operations are a significant drag, but choosing, in general, to disregard rotations for large deformations seems like could lead to erroneous results.

Others can chime in with more accurate explanations.

**Reply by srinath-chakravarthy** [reply]
I agree with @recuero. For any finite strain kinematics rotations have to be done to get the correct cauchy stress in the deformed configuration. That being said if your material is already doing the rotations to get the correct cauchy stress then you can set it to false. For example I have a radial return material that performs this rotation internally because I have a different way of computing the stress in the configuration.  

In short with vanilla moose and all the radial return methods already implemented in tensor mechanics, this should be set to true if you're doing finite strain calculations. 

Cheers
Srinath

**Reply by tophmatthews** [reply]
So perhaps it's my ignorance about what "rotation" means. My mesh is a 2DRz cylinder. Is rotation of the mesh rotating (not really possible), or can it include large deformation?

**Reply by srinath-chakravarthy** [reply]
@tophmatthews i think the rotations here come from the polar decomposition of the deformation gradient and are not the same as a rigid body rotation. Rigid body rotations i believe have to handled independently. Any other mechanics experts, please correct me if I am wrong

**Reply by recuero** [reply]
Yeah, so rigid body rotations happen if F = RU, for U = identity (other issues such as correct BCs would come into play here for most of our static simulations). In general, R (rotation matrix) is needed to capture the local change of orientation. So, it's not so much a mesh rotating, but the result of applying locally (integration point wise) the polar decomposition. You could have drastic changes of orientation without rigid body modes.

**Reply by srinath-chakravarthy** [reply]
Here is a decent example of why we need to do rotations when using the cauchy stress, this is from abaqus theory manual. Note moose uses a variant of the Hughes widget algorithm for computing the strains (Rashid 1993) i think

https://classes.engineering.wustl.edu/2009/spring/mase5513/abaqus/docs/v6.6/books/stm/default.htm?startat=ch03s02ath60.html

**Reply by tophmatthews** [reply]
Okay great, so I need them. I also misread the code, and indeed the stress calculations are consistent due to `_is_elasticity_tensor_guaranteed_isotropic` being true.

**Reply by sapitts** [reply]
@WilkAndy and I discussed this issue in detail a few years back on one of his pull requests (I think on one of the Cosserat PRs), but I haven't been able to find that PR and discussion yet. Once I do I'll post a link here for completeness.

**Reply by WilkAndy** [reply]
Hi @sapitts .  It *might* have been https://github.com/idaholab/moose/pull/9438 .   It's fun to read these old discussions.  Anyway, i concur with what's been written above
